### DISCUSSION TITLE
Slimmer Wiremod

### DISCUSSION THREAD

**Discussion by vurvdev** [wiremod/wire#2916] [Suggestions]
Wiremod is 52MB. It doesn't seem too bad, until you realize that only 4MB of that is the actual code... **<10%** of the download is what matters :\

The largest part is the materials, taking up 33MB, then models, taking up 15MB. (Sounds only take like a megabyte)

I doubt we can do anything automatically about optimizing the models, but that's fine since they take less space than the materials.

I think we could get wiremod much smaller if we cut down on textures and forced them to compressed formats (I think DXT1 would be best for file size?).

There's also a bunch of seemingly duplicate textures? Like `materials/models/cheeze/buttons` vs `materials/models/cheeze/buttons2`? I think models has this same issue with cheeze's stuff?

## Update

After a good amount of discussion on the discord, found a bunch of issues

* Cheez has a model for every single button variant instead of just using a single one with a different material/skin. They're also pretty high fidelity (~30kb each model)
* All of the buttons materials are duplicated for red/blue colors
* Even if they were deduplicated, 170kb for a single *button* material is pretty ridiculous, I have no idea how they managed to get their textures this big as it's supposed to be compressed 512x512 with just a single symbol and simple background..
* Fasteroid's materials aren't compressed and are BGRA8888 which has alpha channel but isn't needed for his texture. This is a much less severe issue compared to all of Cheez's sins tho
* Sound files could be shrunk, but it's only 1.4MB so not really necessary to deal with considering materials take up 30x that.

**Comment by ReallyWeirdCat** [comment]
A separate addon for extra models and materials would be good, leaving only the basic stuff in the main addon

**Comment by vurvdev** [comment]
## Small File Count

A good point that was brought up was the amount of lua files wiremod has.
There's a hard lua file limit, and a bunch of tiny files is more work than a single large file for gmod to download.

Easiest solution appears to be making/using a bundler, at least for the sweps/stools/sents, that just puts them in a single file, and uses the dedicated register functions at runtime (https://wiki.facepunch.com/gmod/scripted_ents.Register and https://wiki.facepunch.com/gmod/weapons.Register).

To deal with debug traces, maybe want to store the code as strings and use CompileString individually.
